### DISCUSSION TITLE
Precision of Floating Point Attributes

### DISCUSSION THREAD

**Discussion by tnakaike** [onnx/onnx#3200] [Ideas]
Hi. As you may know, the precision of floating point attributes can be hurt when a graph is created by using [ONNX python helper](https://github.com/onnx/onnx/blob/master/onnx/helper.py). This can degrade the accuracy of a model because floating-point values calculated at training cannot be passed to ONNX correctly.

The following is a simple example to produce this problem.

```
import onnx
from onnx import helper
from onnx import AttributeProto, TensorProto, GraphProto

kwargs = {}
keys = ['A', 'B', 'C']
values = [2.0, 3.0, 4.0]
kwargs['keys_strings'] = keys
kwargs['values_floats'] = values
kwargs['default_float'] = 1.1

t1 = helper.make_tensor_value_info('T1', TensorProto.STRING, [None, 1])
t2 = helper.make_tensor_value_info('T2', TensorProto.FLOAT, [None, 1])
node = helper.make_node('LabelEncoder', ['T1'], ['T2'], 'LE_1', domain='ai.onnx.ml', **kwargs)
graph = helper.make_graph([node], 'label_encoder', [t1], [t2])
model = helper.make_model(graph)
onnx.save(model, 'label_encoder.onnx')
```

This example maps string values "A", "B", "C" to 2.0, 3.0, and 4.0 respectively. Any other string value is mapped to 1.1 (default_float). However, the precision of the default float value is corrupted in the ONNX file (label_encoder.onnx) as follows.

```
    attribute {
      name: "default_float"
      f: 1.100000023841858
      type: FLOAT
    }
```

This is because the float type in Python is the double type in C/C++ as reported in an issue of protocol buffer (https://github.com/protocolbuffers/protobuf/issues/4569).

Setting an environment variable PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION to python (default is cpp) can fix this problem if the loaded model is used by only Python runtime or compiler. However, this problem still exists when a runtime or compiler is implemented in C/C++ like ONNX Runtime.

Possible solutions are:
1. Implement a new helper using C/C++ protocol buffer implementation
2. Add double attribute to each operator that has float attribute

Is there any good solution?

**Comment by Craigacp** [comment]
As most ML models are trained using 32-bit floats, in practice I wouldn't expect this to be an issue, because they'll be floats when they come out of the C code underneath whatever python library created the model, and then represented as doubles in python (which can represent all 32-bit  floats exactly) and then coerced back into floats when stored in the protobuf. Does this actually cause any skew in real code?
